### ISSUE TITLE
Add shortcut to Run on Appengine instead of Run on Server ...

### ISSUE THREAD

**Issue by elharo** [GoogleCloudPlatform/google-cloud-eclipse#225]
This works, now (with the expanded archetype.)

I'd like to make this a higher level menu item though. However:

http://stackoverflow.com/questions/10504620/add-new-maven-run-configuration-to-menu-in-eclipse

Perhaps we can create a completely new menu item such as "Run Local App Engine" that just invokes the relevant maven command?

Also, debugging this doesn't seem to work (breakpoints not hit). If we go this route we need to figure out how to attach a debugger.

**Comment by rdayal** [comment]
As a note to the initial issue description, we will not delegate to the relevant Maven command; we will invoke the DevAppServer using app-tools-lib-for-java.

**Comment by rdayal** [comment]
Worth considering how this will end up working when we have Flex support.

**Comment by briandealwis** [comment]
[Creating Launch Shortcuts](http://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Fguide%2Fdebug_model.htm)

**Comment by chanseokoh** [comment]
I wonder if this shortcut is just an alias to `Run on Server` (i.e., brings up the same server creation dialog), or if it should always auto-create and start a new server instance on the default 8080 port (or possibly reusing an existing server instance, if any, which I don't think is a good idea).

**Comment by briandealwis** [comment]
That's a good question.  I think it should check if there's an existing server definition with this current project/module.  If none, then we consider it like a new server definition with just this module.

**Comment by chanseokoh** [comment]
> I think it should check if there's an existing server definition with this current project/module.

I guess then this is not so trivial to implement. There can be multiple server instances containing the current project/module. We will have to choose an arbitrary one. Also, maybe we should select a stopped server for simplicity, although I believe we can republish modules?

But then, I think there is too much magic going on with this shortcut. Documenting the behavior of this shortcut will require long explanation. Users might not fully understand it, thinking that the behavior is a bit unpredictable.

Always creating a new instance may simplify things (perhaps by auto-selecting a port, i.e., use 0). Or, we could just bring up the server creation dialog (the same thing the `Run on Server` menu does).

**Comment by briandealwis** [comment]
It's worth some investigation.  Here are some relevant pieces from grovelling through some of the WTP `plugin.xml`s:
- `ServerUIPlugin#runOnServer()` and `org.eclipse.wst.server.ui.actions.RunOnServerAction` are wrappers around `org.eclipse.wst.server.ui.internal.actions.RunOnServerActionDelegate`.  `RunOnServerActionDelegate` seems to be a focal point for launching servers.
- `RunOnServerAction`'s javadoc says:
     > "Run on Server" menu action. Allows the user to select an
     > object, and have automatic server creation, launching, and
     >  the appropriate client to appear.
- there are two commands, `org.eclipse.wst.server.launchShortcut.run` and 
`org.eclipse.wst.server.launchShortcut.debug`, that seem to correspond to the _Run on Server_ and _Debug on Server_ , though there's no obvious handlers. So it may be a placeholder.

**Comment by elharo** [comment]
After this weekend, and struggling a lot with service IDs, I am now convinced that this shortcut is not just the same as Run As > Run On Server. From a user's perspective it should simply launch the currently selected project (or projects) without showing intermediate dialogs. The intermediate dialogs are too confusing and complex, and we want this to be straightforward and simple.

From an implementation perspective this means we should always create a new server configured with only the currently selected project.

As an optimization we might be able to reuse the same server if and only if it has this project and only this project configured.

**Comment by briandealwis** [comment]
I've been exploring how to integrate this into the Run/Debug menus using the launch shortcuts.  I've got it working, but it's not a replacement for the _Launch_ command I implemented in #1070.  There's no way to provide a _ranking_ of applicable shortcuts, so the user gets something like the following:
![screen shot 2016-12-15 at 9 48 26 am](https://cloud.githubusercontent.com/assets/202851/21228411/a7f33722-c2ab-11e6-9af2-b2815e605d4b.PNG)

I'm just cleaning up the patch.
